Optimize SiLU (Swish) op in PyTorch

xiaomengy opened the pull request:
Summary:
Optimize SiLU (Swish) op in PyTorch.

Some benchmark result

input = torch.rand(1024, 32768, dtype=torch.float, device="cpu")
forward: 221ms -> 133ms
backward: 600ms -> 170ms

input = torch.rand(1024, 32768, dtype=torch.double, device="cpu")
forward: 479ms -> 297ms
backward: 1438ms -> 387ms

input = torch.rand(8192, 32768, dtype=torch.float, device="cuda")
forward: 24.34ms -> 9.83ms
backward: 97.05ms -> 29.03ms

input = torch.rand(4096, 32768, dtype=torch.double, device="cuda")
forward: 44.24ms -> 30.15ms
backward: 126.21ms -> 49.68ms

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "SiLU"

Differential Revision: D23093593

BIT-silence commented:
linke https://github.com/pytorch/pytorch/issues/43017

ngimel commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -194,21 +198,64 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
+                  .add_input(input)
+                  .build();
+  silu_stub(iter.device_type(), iter);
+  return std::forward_as_tuple(result, buffer);
```
at least on cuda, since you have to read `x` in backward anyway, recomputing `sigmoid` is pretty cheap (compared to writing it in forward and reading it in backward), so given that it also requires extra memory, maybe you can rewrite it without buffer?

BIT-silence replied:
Yes, this is a good point, actually I am testing this and trying to find a proper way to do that on CUDA.

BIT-silence replied:
From my testing, removing the buffer will make cpu backward slightly slower but forward much faster. For CUDA side, both forward and backward become faster. I have updated the code as well as the benchmark results.

gchanan commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -589,6 +596,111 @@ void glu_backward_kernel(TensorIterator& iter) {
+}
+
+template <typename T>
+void SiLUBackwardContiguousKernel(TensorIterator* it) {
```
why do some kernels take TensorIterators as references and some as points?

BIT-silence replied:
Removed.

ngimel commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
 
 Tensor& silu_out(Tensor& result, const Tensor& self) {
-  return at::mul_out(result, self, at::sigmoid(self));
+  auto iter = TensorIterator::unary_op(result, self);
```
please check that result type is same as self, so that `out` variant of the function properly errors out if user sent a wrong argument

BIT-silence replied:
Done

ngimel commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <ATen/native/Activation.h>
 
+#include <tuple>
```
do you need `tuple` include here?

BIT-silence replied:
Removed.

ngimel commented on test/test_torch.py:
```diff
@@ -16725,20 +16725,32 @@ def test_hardsigmoid(self, device, dtype):
-        inputValues = [-1000, -1, 0, 0.5, 1, 2, 1000]
-        expectedOutput = [0.0000, -0.2689, 0, 0.3112, 0.7312, 1.7616, 1000]
-        precision_4dps = 0.0002
+        input_np = np.random.randn(5, 8)
```
I think it's better to just add silu to TestTorchMathOps (you can easily create np reference lambda for it) and remove this test altogether. If you don't want to do that, you can use `compare_with_numpy` helper function here to handle comparison.

BIT-silence replied:
The issue is silu is a function in torch.nn.functional, not an attribute in tensor. It seems currently TestTorchMathMeta test the methods by getattr(x, opstr). Not sure if there is some proper way to do that.

ngimel replied:
Yeah, ok, nevermind then.

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2659,6 +2659,9 @@
   use_c10_dispatcher: full
   python_module: nn
 
+- func: silu_backward.grad_input(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
```
do you need to expose this overload?

BIT-silence replied:
Removed.

ngimel commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -959,6 +959,13 @@ Tensor infinitely_differentiable_gelu_backward(
 
+Tensor infinitely_differentiable_silu_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
```
yep, without buffer it looks much better :-)

ngimel commented on torch/csrc/api/include/torch/nn/modules/activation.h:
```diff
@@ -567,6 +567,27 @@ class TORCH_API GELUImpl : public torch::nn::Cloneable<GELUImpl> {
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ SiLU ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Applies gelu over a given input.
```
silu

BIT-silence replied:
Done.

ngimel reviewed (commented):
This looks great, thank you!

BIT-silence commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
+    const Tensor& grad_output,
+    const Tensor& input) {
+  Tensor grad_input =
+      at::native::empty_like(input, at::MemoryFormat::Contiguous);
```
Done.

BIT-silence commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -195,20 +199,37 @@ Tensor & celu_(Tensor & self, Scalar alpha) {
 
 Tensor silu(const Tensor& self) {
-  return self * at::sigmoid(self);
+  Tensor result = at::native::empty_like(self, at::MemoryFormat::Contiguous);
```
Done.

ngimel commented on test/test_torch.py:
```diff
@@ -16797,20 +16797,32 @@ def test_hardsigmoid(self, device, dtype):
 
     @dtypes(torch.float, torch.double)
     def test_silu(self, device, dtype):
```
you need @skipIfNoScipy here

BIT-silence replied:
Done.

xiaomengy committed:
Optimize SiLU (Swish) op in PyTorch (#42976)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42976

Optimize SiLU (Swish) op in PyTorch.

Some benchmark result

input = torch.rand(1024, 32768, dtype=torch.float, device="cpu")
forward: 221ms -> 133ms
backward: 600ms -> 170ms

input = torch.rand(1024, 32768, dtype=torch.double, device="cpu")
forward: 479ms -> 297ms
backward: 1438ms -> 387ms

input = torch.rand(8192, 32768, dtype=torch.float, device="cuda")
forward: 24.34ms -> 9.83ms
backward: 97.05ms -> 29.03ms

input = torch.rand(4096, 32768, dtype=torch.double, device="cuda")
forward: 44.24ms -> 30.15ms
backward: 126.21ms -> 49.68ms

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "SiLU"

Reviewed By: houseroad

Differential Revision: D23093593

fbshipit-source-id: ca6129d9fce76b569ac7ea0690dd38eac23c221d
